File: Algorithm/Sorting/Heap Sort/heapSort.py
def heapify(arr, n , l):
    
    largest_index = l

    ln = 2*l+1
    rn = 2*l+2

    if ln < n and arr[ln] > arr[l]:
        l = ln
    if rn < n and arr[rn] > arr[l]:
        l = rn
    if l!= largest_index:
        arr[l], arr[largest_index] = arr[largest_index], arr[l]
        heapify(arr, n, l)

def heapSort(arr):

    n = len(arr)
    for i in range(n//2-1, -1, -1):
        heapify(arr, n, i)
    
    for n in range(n-1, 0, -1):
        arr[0], arr[n] = arr[n], arr[0]
        heapify(arr, n, 0)

arr = [12, 11, 13, 5, 6, 7]
heapSort(arr)

print(arr)

File: Algorithm/Sorting/Heap Sort/heapSort.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    void swap(int *a, int *b)
    {
        int temp = *a;
        *a = *b;
        *b = temp;
    }

public:
    void heapify(int arr[], int n, int i)
    {
        int largest_index = i,
            ln = 2 * i + 1,
            rn = 2 * i + 2;

        if (ln < n && arr[ln] > arr[i])
            i = ln;
        if (rn < n && arr[rn] > arr[i])
            i = rn;
        if (i != largest_index)
        {
            swap(&arr[i], &arr[largest_index]);
            heapify(arr, n, i);
        }
    }

public:
    void heapSort(int arr[], int arr_size)
    {
        for (int i = arr_size / 2 - 1; i >= 0; i--)
            heapify(arr, arr_size, i);

        for (int n = arr_size - 1; n > 0; n--)
        {
            swap(&arr[0], &arr[n]);
            heapify(arr, n, 0);
        }
    }
};

int main(int argc, char const *argv[])
{
    int arr[] = {12, 11, 13, 5, 6, 7};
    int arr_size = sizeof(arr) / sizeof(arr[0]);

    Solution().heapSort(arr, arr_size);
    for (auto &&i : arr)
        cout << i << " ";

    return 0;
}
